### PR TITLE
Store session IDs in the account table

### PR DESCRIPTION
**Pull request by jeffpw-goog**
### Description

Please explain the changes you made here.

#### General

Read the full guidelines for PRs [here](https://github.com/civiform/civiform/wiki/Technical-contribution-guide#creating-a-pull-request)

- [ ] Added the correct label: < feature | enhancement | bug | under-development | dependencies | infrastructure | ignore-for-release | database >
- [ ] Assigned to a specific person, `civiform/developers`, or a [more specific round-robin list](https://github.com/civiform/civiform/wiki/Technical-contribution-guide#adding-reviewers)
- [ ] Added an additional reviewer from outside your organization as FYI (if the primary reviewer is in the same organization as you)
- [ ] Removed the release notes section if the title is sufficient for the release notes description, or put more details in that section.
- [ ] Created unit and/or browser tests which fail without the change (if possible)
- [ ] Performed manual testing (Chrome and Firefox if it includes front-end changes)
- [ ] Extended the README / documentation, if necessary. For user-facing features, consider updating [the user docs](https://github.com/civiform/docs). For "under-the-hood" changes or things more relevant to developers, consider updating [the dev wiki](https://github.com/civiform/civiform/wiki).

#### Database evolutions

Read the guidelines [here](https://github.com/civiform/civiform/wiki/Database#writing-database-evolutions)

- [ ] Assigned two reviewers
- [ ] Guarded against already existing resources using `IF NOT EXISTS` and `IF EXISTS`
- [ ] Downs created to undo changes in Ups
- [ ] Data migrations aren't being done (please use a [Durable Job](https://github.com/civiform/civiform/wiki/Database#durable-jobs-for-data-updates) if doing a data migration)

#### Durable jobs

Read the docs [here](https://github.com/civiform/civiform/wiki/Database#durable-jobs-for-data-updates)

- [ ] Assigned two reviewers
- [ ] Included a rollback plan and a job to undo the data changes if appropriate

#### User visible changes

- [ ] Followed steps to [internationalize](https://github.com/civiform/civiform/wiki/Internationalization-%28i18n%29#application-strings) any new strings
  - [ ] Added context strings to new [messages](https://github.com/civiform/civiform/blob/main/server/conf/i18n/messages)
  - [ ] Didn't use a message in applicant facing code that isn't translated yet (unless behind a flag)
- [ ] Wrote browser tests using the [validateAccessibility](https://sourcegraph.com/github.com/civiform/civiform/-/blob/browser-test/src/support/index.ts?L437:14&subtree=true) method
- [ ] Made equivalent changes in Thymeleaf for applicant-facing features and changes (or created an issue in the [Northstar epic](https://github.com/orgs/civiform/projects/1/views/94) to track that it's missing from Thymeleaf)
- [ ] Tested on mobile view. See [mobile device mode](https://developer.chrome.com/docs/devtools/device-mode/)
- [ ] Manually tested at 200% size
- [ ] Manually evaluated tab order
- [ ] Manually tested with a screen reader if the feature is applicant-facing. See [screen reader testing](https://github.com/civiform/civiform/wiki/Testing#screen-reader-testing)

#### New Features

- [ ] Add new FeatureFlag env vars to `server/conf/helper/feature-flags.conf`
- [ ] Conditioned new functionality on a [FeatureFlag](https://github.com/civiform/civiform/wiki/Feature-Flags)
- [ ] Wrote browser tests with the feature flag off and on, etc.

### Instructions for manual testing

If instructions are needed for manual testing by reviewers, include them here.

### Issue(s) this completes

Fixes #<issue_number>; Fixes #<issue_number>...

### CODE REVIEW

File: server/app/repository/AccountRepository.java [hunk 1]
```diff
@@ -469,6 +470,11 @@ public void addIdTokenAndPrune(AccountModel account, String sessionId, String id
     }
     idTokens.purgeExpiredIdTokens(clock);
     idTokens.storeIdToken(sessionId, idToken);
+
+    // For now, we set the duration to Auth0s default of 10 hours.
```
Comment (dkatzz) [inline server/app/repository/AccountRepository.java:474]: **thought:** I wonder if this should be behind the flag as well?

File: server/app/repository/AccountRepository.java [hunk 2]
```diff
@@ -469,6 +474,13 @@ public void addIdTokenAndPrune(AccountModel account, String sessionId, String id
     }
     idTokens.purgeExpiredIdTokens(clock);
     idTokens.storeIdToken(sessionId, idToken);
+
+    if (settingsManifest.getSessionReplayProtectionEnabled()) {
+      // For now, we set the duration to Auth0s default of 10 hours.
+      account.removeExpiredActiveSessions(clock, Duration.ofHours(10));
```
[the remark this answers is not in the document]
Comment (gwendolyngoetz) [reply server/app/repository/AccountRepository.java:480]: (thought) It might be more work, but rather than pushing the clock and the duration as separate things a new class tailored to the session lifecycle would go a long way to making the intent and tests easier to read.

```psueudocode
class SessionLifecycle {
  ctor(Clock) {}
  Duration maxSessionDuration() {}
  ...
}
```

File: server/test/repository/AccountRepositoryTest.java [hunk 3]
```diff
@@ -350,9 +361,16 @@ public void deleteUnusedGuestAccounts() {
 
   @Test
   public void addIdTokenAndPrune() {
+    when(mockSettingsManifest.getSessionReplayProtectionEnabled()).thenReturn(false);
     AccountModel account = new AccountModel();
     String fakeEmail = "fake email";
     account.setEmailAddress(fakeEmail);
+    // Clock that is within the default session duration
+    Clock clock = Clock.fixed(Instant.now().minusSeconds(10000), ZoneId.systemDefault());
```
Comment (gwendolyngoetz) [inline server/test/repository/AccountRepositoryTest.java:369]: (blocking) It's a bit hard to follow what makes the time valid/invalid without knowing the duration. It would increase the readability if instead of `clock` calling it `validSessionClock` or something like that. here and on the other clocks.

File: server/test/repository/AccountRepositoryTest.java [hunk 4]
```diff
@@ -378,6 +396,66 @@ public void addIdTokenAndPrune() {
...
   }
 
+  @Test
+  public void addIdTokenAndPrune_sessionReplayEnabled() {
+    when(mockSettingsManifest.getSessionReplayProtectionEnabled()).thenReturn(true);
+    AccountModel account = new AccountModel();
+    String fakeEmail = "fake email";
+    account.setEmailAddress(fakeEmail);
+    // Clock that is within the default session duration
+    Clock clock = Clock.fixed(Instant.now().minusSeconds(10000), ZoneId.systemDefault());
```
Comment (gwendolyngoetz) [inline server/test/repository/AccountRepositoryTest.java:406]: See previous comment

File: server/test/repository/AccountRepositoryTest.java [hunk 5]
```diff
@@ -378,6 +396,66 @@ public void addIdTokenAndPrune() {
...
+  }
+
+  @Test
+  public void addIdTokenAndPrune_throwsWithoutActiveSession() {
+    when(mockSettingsManifest.getSessionReplayProtectionEnabled()).thenReturn(true);
+    AccountModel account = new AccountModel();
+    String fakeEmail = "fake email";
+    account.setEmailAddress(fakeEmail);
+    // Clock is above the default session duration
+    Clock clock = Clock.fixed(Instant.now().minusSeconds(37000), ZoneId.systemDefault());
```
Comment (gwendolyngoetz) [inline server/test/repository/AccountRepositoryTest.java:443]: See previous comment

File: server/test/repository/AccountRepositoryTest.java (hunk 3 above)
Comment (dkatzz) [reply server/test/repository/AccountRepositoryTest.java]: Good call, updated!

File: server/app/repository/AccountRepository.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (dkatzz) [reply server/app/repository/AccountRepository.java]: Updated to create that class! And also agree around session duration being configurable - created a bug to track and added a TODO

File: server/app/models/SessionLifecycle.java [hunk 6]
```diff
@@ -0,0 +1,22 @@
+package models;
+
+import java.time.Clock;
+import java.time.Duration;
+
+public final class SessionLifecycle {
```
Comment (gwendolyngoetz) [inline server/app/models/SessionLifecycle.java:6]: (blocking) Can you add javadoc?

File: server/test/repository/AccountRepositoryTest.java [hunk 7]
```diff
@@ -23,19 +27,32 @@
...
 public class AccountRepositoryTest extends ResetPostgres {
   public static final String EMAIL = "email@email.com";
   public static final String PROGRAM_NAME = "program";
   public static final String AUTHORITY_ID = "I'm an authority ID";
+  // Clock that is within the default session duration (less than 10 hours)
+  public static final Clock VALID_SESSION_CLOCK =
+      Clock.fixed(Instant.now().minusSeconds(10), ZoneId.systemDefault());
+  // Clock is outside of the default session duration (greater than 10 hours / 36000 seconds)
+  public static final Clock INVALID_SESSION_CLOCK =
+      Clock.fixed(Instant.now().minusSeconds(37000), ZoneId.systemDefault());
```
Comment (gwendolyngoetz) [inline server/test/repository/AccountRepositoryTest.java:43]: (praise) This help so much! Thank you!

**Review by gwendolyngoetz — CHANGES_REQUESTED** [review]

File: server/test/repository/AccountRepositoryTest.java (hunk 7 above)
Comment (dkatzz) [reply server/test/repository/AccountRepositoryTest.java:43]: Agreed! Thank you for the suggestion :)

File: server/app/models/SessionLifecycle.java (hunk 6 above)
Comment (dkatzz) [reply server/app/models/SessionLifecycle.java:7]: Added!

**Review by gwendolyngoetz — APPROVED** [review]
